src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+            safeContentsAsn.ContentType = DecryptedSentinel;
+        }
+
+        private static void ProcessSafeContents(
+            in ContentInfoAsn safeContentsAsn,
+            ref CertBagAsn[] certBags,
+            ref AttributeAsn[][] certBagAttrs,
+            ref int certBagIdx,
+            ref SafeBagAsn[] keyBags,
+            ref int keyBagIdx)
+        {
+            ReadOnlyMemory<byte> contentData = safeContentsAsn.Content;
+
+            if (safeContentsAsn.ContentType == Oids.Pkcs7Data)
+            {
+                contentData = Helpers.DecodeOctetStringAsMemory(contentData);
+            }
+
+            AsnReader outer = new AsnReader(contentData, AsnEncodingRules.BER);
+            AsnReader reader = outer.ReadSequence();
+            outer.ThrowIfNotEmpty();
+
+            while (reader.HasData)
+            {
+                SafeBagAsn.Decode(reader, out SafeBagAsn bag);
+
+                if (bag.BagId == Oids.Pkcs12CertBag)
+                {
+                    CertBagAsn certBag = CertBagAsn.Decode(bag.BagValue, AsnEncodingRules.BER);
+
+                    if (certBag.CertId == Oids.Pkcs12X509CertBagType)
+                    {
+                        GrowIfNeeded(ref certBags, certBagIdx);
+                        GrowIfNeeded(ref certBagAttrs, certBagIdx);
+                        certBags[certBagIdx] = certBag;
+                        certBagAttrs[certBagIdx] = bag.BagAttributes;
+                        certBagIdx++;
+                    }
+                }
+                else if (bag.BagId == Oids.Pkcs12KeyBag || bag.BagId == Oids.Pkcs12ShroudedKeyBag)
+                {
+                    GrowIfNeeded(ref keyBags, keyBagIdx);
+                    keyBags[keyBagIdx] = bag;
+                    keyBagIdx++;
+                }
+            }
+        }
+
+        private AsymmetricAlgorithm LoadKey(SafeBagAsn safeBag, ReadOnlySpan<char> password)
+        {
+            if (safeBag.BagId == Oids.Pkcs12ShroudedKeyBag)
+            {
+                ArraySegment<byte> decrypted = KeyFormatHelper.DecryptPkcs8(
+                    password,
+                    safeBag.BagValue,
+                    out int localRead);
+
+                try
+                {
+                    if (localRead != safeBag.BagValue.Length)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+
+                    return LoadKey(decrypted.AsMemory());
+                }
+                finally
+                {
+                    CryptoPool.Return(decrypted.Array, clearSize: decrypted.Count);
+                }
+            }
+
+            Debug.Assert(safeBag.BagId == Oids.Pkcs12KeyBag);
+            return LoadKey(safeBag.BagValue);
+        }
+
+        private static void GrowIfNeeded<T>(ref T[] array, int idx)
+        {
+            T[] oldRent = array;
```

Comment:
Inside the if?

Reply:
Went the other way: This read and the later assignment should have been the only two places that used the ref variable.